WebDriverAgentLib/Categories/XCUIElement+FBFind.m

```diff
@@ -69,51 +80,87 @@ - (void)descendantsWithProperty:(NSString *)property value:(NSString *)value par
 
 #pragma mark - Search by Predicate String
 
-- (NSArray<XCUIElement *> *)fb_descendantsMatchingPredicate:(NSPredicate *)predicate {
+- (NSArray<XCUIElement *> *)fb_descendantsMatchingPredicate:(NSPredicate *)predicate shouldReturnAfterFirstMatch:(BOOL)shouldReturnAfterFirstMatch
+{
   XCUIElementQuery *query = [[self descendantsMatchingType:XCUIElementTypeAny] matchingPredicate:predicate];
-  NSArray *childElements = [query allElementsBoundByIndex];
-  return childElements;
+  NSMutableArray *result = [NSMutableArray array];
+  if (shouldReturnAfterFirstMatch) {
+    XCUIElement *match = [query elementBoundByIndex:0];
+    if (match && [match exists]) {
+      [result addObject:match];
+    }
+  } else {
+    [result addObjectsFromArray:[query allElementsBoundByIndex]];
+  }
+  return result.copy;
 }
 
 
 #pragma mark - Search by xpath
 
-- (NSArray<XCUIElement *> *)fb_descendantsMatchingXPathQuery:(NSString *)xpathQuery
+- (NSArray<XCElementSnapshot *> *)getMatchedSnapshotsByXPathQuery:(NSString *)xpathQuery
 {
   // XPath will try to match elements only class name, so requesting elements by XCUIElementTypeAny will not work. We should use '*' instead.
   xpathQuery = [xpathQuery stringByReplacingOccurrencesOfString:@"XCUIElementTypeAny" withString:@"*"];
-  NSArray *matchingSnapshots = [self.lastSnapshot fb_descendantsMatchingXPathQuery:xpathQuery];
-  NSArray *allElements = [[self descendantsMatchingType:XCUIElementTypeAny] allElementsBoundByIndex];
-  NSArray *matchingElements = [self filterElements:allElements matchingSnapshots:matchingSnapshots];
+  return [self.lastSnapshot fb_descendantsMatchingXPathQuery:xpathQuery];
+}
+
+- (NSArray<XCUIElement *> *)fb_descendantsMatchingXPathQuery:(NSString *)xpathQuery shouldReturnAfterFirstMatch:(BOOL)shouldReturnAfterFirstMatch
+{
+  NSArray *matchingSnapshots = [self getMatchedSnapshotsByXPathQuery:xpathQuery];
+  if (0 == [matchingSnapshots count]) {
+    return @[];
+  }
+  if (shouldReturnAfterFirstMatch) {
+    matchingSnapshots = @[[matchingSnapshots firstObject]];
```

Comment:
yes, exactly. xpath evaluation with libxml is super fast and it takes milliseconds to get either one or multiple matches from there. But then we have _filterElements_ call, which enumerates all the found snapshots and matches them to the actual elements, which is super slow. By decreasing count of actual snapshots to one we can skip all the further _snapshots_  array iterations there, which speeds up the search operation significantly.